### ISSUE TITLE
t.internet.process.Process should check for valid arguments

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10531]
|<img alt="marienz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| marienz reported|
|-|-|
|Trac ID|trac#815|
|Type|enhancement|
|Created|2004-12-08 19:39:50Z|

Attachments:

* [process.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d52/d528edaa45e66e08a9ece98272130b42e77cef55/9ea0ea20446ecb340c51da47d5a9dcce673ca7e5.patch) (1914 bytes) - added by marienz on 2004-12-08 19:39:47Z - 
* [process_815.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d52/d528edaa45e66e08a9ece98272130b42e77cef55/fe99503e431d5e853929fadcfd01a7a98ca858c6.py) (2230 bytes) - added by therve on 2006-10-02 13:23:26Z - 
* [process_815_2.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d52/d528edaa45e66e08a9ece98272130b42e77cef55/33323e757fe88fd72b366580de539c49acf184d6.py) (2910 bytes) - added by therve on 2006-10-02 14:13:14Z - 
* [process_815_3.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d52/d528edaa45e66e08a9ece98272130b42e77cef55/6a204b9e04957d1275c8bfc4d5abdbc9740690ee.py) (3714 bytes) - added by therve on 2006-10-30 10:20:20Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__815 815
type__enhancement enhancement
reporter__marienz marienz
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1102534790000000 1102534790000000
changetime__1166061414000000 1166061414000000
version__ 
owner__ 
cc__exarkun cc__itamarst cc__marienz cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="marienz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|marienz commented|
|-|-|

```
#!html
<pre>
If you pass Process arguments that don't make sense you may not find out until
it tries to call os.execvpe. Unfortunately that call happens after forking and
playing interesting games with fd's, so the exception can "vanish". 

The following patch tries to do some sanity checking before forking.
Unfortunately the actual checks execvpe does are in C code, and I can't
duplicate it exactly. So this won't catch every bad thing you can throw at exec,
and it will complain (rarely) about valid things. See the comments for details.

This does cause two new test failures, both in mail.test.test_mail.py,
TestAliasResolution and TestCyclicAlias. But those tests seem to be a bit bogus
(afaik) so I don't think the failures are a bad thing :)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

Attach a patch with small modification from previous patch (use TypeError, add a new check for string arguments, remove check for empty arguments), and a testcase.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

glanced at the patch briefly; one other thing that will definitely cause exec to fail is a string with nul bytes in it.  if we are going to do validation, we should catch and reject these too.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

Added the null bytes test.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_6"></a>|@ralphm commented|
|-|-|

therve, it appears to me that you are not going to review this yourself. It might be useful to reassign ownership.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve removed owner|
|-|-|

Ups, thank you.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

Committed this patch to spawnprocess-args-815 with some minor edits.  I think it's good to merge, but someone else should take a quick look.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Oops, I can't review this one.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

On IRC Jerub told me that he reviewed this and it was okay to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [18491]) Merge spawnprocess-args-815

Author: therve, exarkun
Reviewer: jerub
Fixes [#10531](https://github.com/twisted/twisted/issues/10531)

Perform typechecking in the POSIX process launching code and raise TypeErrors
before forking if the provided arguments cannot result in a successful execvpe()
call.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `reopened`|
|-|-|

Re-opened by r18491

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set owner to @therve<br>@therve set status to `new`|
|-|-|

Attached a patch moving the check into posixbase, and using it into win32reactor. It **seems** to be good under Windows.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to moonfallen|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

I was just taking another look at this and noticed that spawnProcess lets unicode strings through.  This only works if the unicode string can be encoded using the 'ascii' codec, so it's not clear that we should actually be allowing it.  At the very least there should be test coverage for this behavior.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

I ended up keeping unicode support, but deprecating it.  I need to test this more.  The Windows buildslaves are offline right now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I guess this seems to be working now.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_22"></a>|@jml commented|
|-|-|

Branch is now `spawnprocess-args-815-3`

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_23"></a>|@jml set owner to @exarkun|
|-|-|

Looks good. I particularly like the way you've defined the utility classes for test_process.

 * `_deprecatedUnicodeSupportTest` should really have a docstring. Given that it has assertions, it should explain what it's checking for in its docstring.
 * The docstring for `_checkProcessArgs` should say the `env` part of the two-tuple is either a dict or `None`. Or, it should refer to another docstring which explains the return values.

Fix these and merge away.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun set status to `assigned`|
|-|-|

Fixed those.  pre-checkin testing found one problem with gtk2 reactor :/  I'm not sure how to best fix it yet.  gtk2 sets the default system encoding to 'utf-8' which breaks assumptions made by the tests.  I thought the fix for this was to use 'ascii' instead of getdefaultencoding but now I think that's wrong, and it should still use getdefaultencoding but the tests should be different somehow.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun set owner to @jml<br>@exarkun set status to `new`|
|-|-|

I've committed some more changes and a few comments which hopefully make it very clear what is supposed to be going on and what is going on.  I'll leave them to explain the changes I made.  If they're not clear, please let me know and I'll have another go at them.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_26"></a>|@jml set owner to @exarkun|
|-|-|

Awesome.

Only change required is from "Anything unicode string" to "Any unicode string" in posixbase.py. Then merge away. Thanks for making this all so clear.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun set status to `closed`|
|-|-|

(In [18898]) Merge spawnprocess-args-815-3

Author: therve, exarkun
Reviewer: exarkun, jml
Fixes [#10531](https://github.com/twisted/twisted/issues/10531)

Check the argv and environment values before creating a child process with
IReactorProcess.spawnProcess and reject values which will not be allowed
by the actual process creation API.  This causes spawnProcess to raise
synchronously, rather than delivering bytes on stderr to the process
protocol used.

Reject anything which is neither str nor unicode for both of these things.
Also emit a deprecation warning when unicode is received, since this really
shouldn't have been allowed at all.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
